content/blog/2019-09-25-manually-managed-memory.md

```diff
@@ -0,0 +1,303 @@
+other words, programs may `alloc` memory that they never `free`.
+
+For a larger example of how pointers can be used in Bril, the following C code:
+```C
+int* vals = malloc(sizeof(int)*10);
+vals[0] = 0;
+for (int i = 1; i < 10; i++) {
+  vals[i] = vals[i-1] + 4;
+}
+```
+
+Would be roughly equivalent to the following Bril code:
+```C
+ ten: int = const 10;
+ zero: int = const 0;
+ one: int = const 1;
+ neg_one: int = const -1;
+ four: int = const 4;
+ vals: ptr<int> = alloc ten
+ store vals zero;
+ i: int = const 1;
+ i_minus_one: int = add i neg_one;
+loop:
+ cond: lt i ten;
+ br cond done body;
+body:
+ vals_i: ptr<int> = ptradd vals i;
+ vals_i_minus_one: ptr<int> = ptradd vals i_minus_one;
+ tmp: int = load vals_i_minus_one;
+ tmp: int = add tmp four;
+ store vals_i tmp;
+ i = add i one;
+ i_minus_one = add i_minus_one one;
+ jmp loop;
+done:
+ free vals;
+ ret;
+```
+
+### Implementation
+
+We implemented our design by extending the Bril text parser to support
+pointer types and the Bril interpreter to support pointers,
+heap memory, and runtime error checking.
+
+#### Pointer Representation
+
+A Bril pointer is represented by a pointer with two fields: a *Key* that points
+into the [*Heap*](#heap-memory), and a *tag* that tells the runtime which kind
+of data the memory cell should be used for. Type tags can be `"int"`, `"bool"`
+or `"ptr"`. Type tags are checked whenever a Bril program does a `store` to make
+sure that the cells of the allocated memory were allocated to store something of
+that type. All pointers, regardless of what they point to, have the same type
+tag. This is because all pointers have the same in-memory representation, so
+ensuring a cell for pointers is only ever used for pointers can be done without
+worrying about what type of data the pointer points to.
+
+#### Heap Memory
+
+Memory itself is represented by the `Heap` data structure. It implements all the
+operations exposed to Bril programs. Concretely, the `Heap` implementation is
+a Javascript `Map` that maps `number` keys to arrays of objects. Its `alloc`
+method extends the map to map a fresh key `number` to a new array of the
+requested size and returns an opaque *Key* wrapping the `number`.
+
+A *Key* is a pair of a `base` and an `offset`. The `base` is the key used to
+look up the array in the heap `Map` and the `offset` is an index into the array.
+Freshly allocated pointers have `offset == 0`, and pointer arithmetic can only
+change the `offset`. In this way, we keep track of which allocation any given
+pointer belongs to, regardless of "where" it may point in memory. Notably this
+implementation of a Heap does not model "a single contiguous memory space"; each
+allocation represents a continguous space and allocations are otherwise
+unrelated.
+
+The `free` method deletes entries from the internal map, so we are relying on
+the base Typescript Map implementation and the Javascript runtime garbage
```

Comment:
As you noted above, `Map` is a [JavaScript standard library thing](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Map) rather than a TypeScript thing per se.